Fix #2223: explorations now appear sorted in Top Rated and Recently Published pages

Arunabh98 opened the pull request:
After discussion in #2223 we decided to fetch the explorations for top rated and recently published pages by database query. Also I did not have to create a additional handler to take care of these two cases. Instead I extended the functionality of the current handler to take care of these two cases. I extended the url schema so that we can take care of these two cases without changing the urls of the current pages. I also changed the database query function so that we can fetch a variable number of explorations which is necessary as we have to show different number of explorations in the main library page and those two pages. Made some changes to LibraryPage handler in library.py to account for changes in urls and pass title from the backend itself.

@seanlip

seanlip commented on core/controllers/library.py:
```diff
@@ -93,48 +108,78 @@ def get(self):
     """Provides data for the default library index page."""
 
-    def get(self):
+    def get(self, **kwargs):
```
No, this handler should be unchanged, I think. It's for the library index page. If we need separate JSON handlers for the other pages let's just make new JSON handlers that each have one task, otherwise there's going to be too much cyclomatic complexity in this one since it has to do so many different things.

Arunabh98 replied:
Made a separate handler...

seanlip commented on core/controllers/library.py:
```diff
@@ -70,9 +70,22 @@ class LibraryPage(base.BaseHandler):
     """
 
-    def get(self):
+    def get(self, **kwargs):
```
Don't use kwargs -- this is a code smell, it's too 'open'. Instead of overloading this handler, how about making a new handler instead for the other pages -- say, LibraryGroupPage or something?

Arunabh98 replied:
Removed usage kwargs. I will keep this in mind next time!

seanlip commented:
Ah, just read your top comment. I do think it is better to create an additional handler, and if these handlers use lots of common functionality, to factor out that common functionality. The current way introduces far too much complexity in the controller layer, which is supposed to be a thin lightweight bridge between the URL endpoints and the domain logic.

Arunabh98 commented:
Okay I'll create separate handlers for these two pages. That won't take long. Though I have one problem. Using optional arguments I was able to eliminate the need to have a conditional HTTP get requests in `Library.js`. Currently I have only one http get request which can handle all cases. Now if I have separate handlers I will have to have a conditional get request - one for these two pages and other for the main library page (and this is bad as there is a coupling between backend and frontend). However I can avoid this if I am able to send the url to ping from the backend. Can you guide me in a suitable direction as how will I achieve this?

seanlip commented:
Hm, so, I got an email notification of a comment regarding a conditional get request, but that comment seems to have disappeared. That said, I'm guessing you did indeed find a way to send the backend URL to the frontend (which ought to be straightforward)?

Arunabh98 commented:
I am still not able to figure out how to access the variable passed in library.py in Library.js. For example, I pass a variable in `LibraryPage` handler called `full_results_url`. However I am not able to access the variable in Library.js. Could you give an idea on how to achieve this?

(I had deleted the comment because I thought I had figured it out)

seanlip commented:
Wouldn't you be passing it in LibraryIndexHandler when you send the definitions for each group to the frontend? I thought it would live inside each of the group dicts (i.e. next to the keys activity_summary_dicts, categories, header) and be accessed in similar fashion to how those other keys are accessed.

Arunabh98 commented:
Oh I was thinking of something else. I have passed the urls from the backend. If the url passed is None, a search query is performed.

Arunabh98 commented:
Hi, I think this is ready for another review!

seanlip commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+                summary_services.get_recently_published_exploration_summary_dicts( # pylint: disable=line-too-long
+                    feconf.RECENTLY_PUBLISHED_QUERY_LIMIT_FULL_PAGE))
+            if recently_published_summary_dicts:
+                summary_dicts_by_category.insert(0, {
```
You could just append or say `= [...]`? I mean, there's at most one, and the variable is an empty list before that.

Ditto below.

Arunabh98 replied:
Done

seanlip commented on core/domain/exp_services.py:
```diff
@@ -366,20 +366,23 @@ def get_non_private_exploration_summaries():
 
 
-def get_top_rated_exploration_summaries():
+def get_top_rated_exploration_summaries(number_of_explorations_to_fetch):
```
Let's call this argument name 'limit'. Shorter, and consistent with usages elsewhere in the codebase.

Arunabh98 replied:
Done

seanlip commented on core/domain/summary_services.py:
```diff
@@ -415,32 +416,36 @@ def get_featured_activity_summary_dicts(language_codes):
         filtered_exp_summaries,
         key=lambda exp_summary: exp_summary.scaled_average_rating,
-        reverse=True)[:feconf.NUMBER_OF_TOP_RATED_EXPLORATIONS]
+        reverse=True)[:number_of_explorations_to_fetch]
```
But you've already filtered to a max limit above, right -- so this is unnecessary?

Arunabh98 replied:
Removed

seanlip commented on core/templates/dev/head/library/library.html:
```diff
@@ -171,7 +187,7 @@ <h2 class="oppia-library-group-header">
 {% block footer %}
   <div ng-controller="LibraryFooter">
-    <div ng-if="!inSearchMode">
+    <div ng-if="!inSearchMode && !inRankMode">
```
Is rank-mode infinite scroll? If not, you can leave the footer around in that case.

Arunabh98 replied:
Done

seanlip commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+        """Handles GET requests."""
+
+        search_mode = 'search' in self.request.url
+        rank_mode = 'explorations' in self.request.url
```
Why needed? It's not read by the frontend.

That said, in this page, it's obviously True...

Arunabh98 replied:
Removed

seanlip commented on core/templates/dev/head/library/Library.js:
```diff
@@ -40,34 +40,72 @@ oppia.controller('Library', [
+        }
+      });
+    } else {
+      $http.get('/libraryindexhandler').success(
```
Oh, _interesting_ -- I just realized that this handler gets called way more than it needs to be. It even gets called during the search results, when the library groups are not actually showing. Perhaps we should put "else if (!$scope.searchMode)" here.

Arunabh98 replied:
Done

seanlip commented on core/templates/dev/head/library/Library.js:
```diff
@@ -40,34 +40,72 @@ oppia.controller('Library', [
+
+        $rootScope.loadingMessage = '';
+
+        // Pause is necessary to ensure all elements have loaded, same for
```
Why are we talking about carousels here? Is any of this actually needed?

Arunabh98 replied:
Removed

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -73,6 +73,12 @@ class LibraryPage(base.BaseHandler):
         search_mode = 'search' in self.request.url
+        rank_mode = 'explorations' in self.request.url
+
+        if search_mode:
```
Right now I think the {{ title }} will automatically be translated once I make a i18n variables for the strings. For example the title for the main library page will automatically be translated because I am passing a i18n variable for it. 
Look at this comment in `base.html` on line 62:
```
<!-- The title is bound to the rootScope. The content of the block
maintitle can be a string or a translation id. If it is a translation it
will be replaced by its translation when the page is loading. If it is a
string it would be displayed as is. This is the only way to translate
the page title because the head of the file is outside the scope of
any other controller. -->
```
So I think to solve this problem we have to translate the strings and make i18n variables for them (which can be done in a seperate PR). Is it okay if I pass the title directly from the backend?
What are your views on this?

seanlip replied:
Well, I was thinking of making the 'minimal' change, which implies leaving it in the frontend, for no other reason than that it's currently in the frontend. At some point, it seems reasonable to pass the title (or, rather, the title i18n code) from the backend, just as we do for the group headers, but when we make that change I think we should do it fully and properly. Hence the suggestion to postpone it for now.

Arunabh98 replied:
Oh okay

Arunabh98 replied:
So for now do I leave the title to be 'Find explorations to learn from - Oppia' for the top rated and recently published pages?

seanlip replied:
Yes, I think that is fine.

Arunabh98 replied:
Done

Arunabh98 commented:
How should I modify these tests:
```
        response = self.testapp.post('/library/extra', {}, expect_errors=True)
        self.assertEqual(response.status_int, 404)

        response = self.testapp.put('/library/extra', {}, expect_errors=True)
        self.assertEqual(response.status_int, 404)
```
in `base_test.py` in `test_requests_for_invalid_paths` on line 83? These tests fails because I have `/library/<group_name>` in main.py.

seanlip commented:
Ah -- oops, I should have caught that. The handler should check the value of rank_mode (or whatever we're calling it). If it's invalid, it should 404.

Arunabh98 commented:
I have already introduced the tests like:
```
        if (group_name == feconf.LIBRARY_GROUP_RECENTLY_PUBLISHED or
                group_name == feconf.LIBRARY_GROUP_TOP_RATED):
            group_mode = True
        else:
            raise self.PageNotFoundException
```
However I think those two functions are testing POST and PUT requests which LibraryGroupPage does not have. These two tests are passing:
```
        response = self.testapp.get('/library/extra', expect_errors=True)
        self.assertEqual(response.status_int, 404)

        response = self.testapp.get('/library/data/extra', expect_errors=True)
        self.assertEqual(response.status_int, 404)
```
But the above two tests are passing? I am getting an error like this:
```
ERROR:root:Exception raised: post() got an unexpected keyword argument 'group_name'
```
So do I make post and put functions and then return page not found exception? Or do I remove these two tests? Or do you suggest something else?

seanlip commented:
Offhand, I don't know, I'm afraid. Needs a bit of digging -- see what other handlers do for `<exploration_id>`, for example. I'm pretty sure some of them have, e.g., only a put() method and everything else is undefined. Why do the tests fail only in this case, and not for those handlers?

Arunabh98 commented:
So I think I have figured out the problem. The problem arises due to lines 265-279 in `core/controllers/base.py`:
```
    def get(self, *args, **kwargs):  # pylint: disable=unused-argument
        """Base method to handle GET requests."""
        raise self.PageNotFoundException

    def post(self, *args):  # pylint: disable=unused-argument
        """Base method to handle POST requests."""
        raise self.PageNotFoundException

    def put(self, *args):  # pylint: disable=unused-argument
        """Base method to handle PUT requests."""
        raise self.PageNotFoundException
```
As we can see the the `**kwargs` is not included for post and put requests due to which it returns 500 error and raises an error like: 
```
ERROR:root:Exception raised: post() got an unexpected keyword argument 'group_name'
```
On changing the code to:
```
    def get(self, *args, **kwargs):  # pylint: disable=unused-argument
        """Base method to handle GET requests."""
        raise self.PageNotFoundException

    def post(self, *args, **kwargs):  # pylint: disable=unused-argument
        """Base method to handle POST requests."""
        raise self.PageNotFoundException

    def put(self, *args, **kwargs):  # pylint: disable=unused-argument
        """Base method to handle PUT requests."""
        raise self.PageNotFoundException
```
these problems get solved. So do I proceed with this solution?

seanlip commented:
Good find!

I suggest something different as a fix, though. I think the issue arises because we are using `<group_name>` in the URL. This is actually a bit non-standard since we generally only ever do that for `<exploration_id>` and use a decorator to handle it. So, to keep things consistent here, can we just define two separate URLs and two separate handlers, since these are essentially treated as separate pages?

Arunabh98 commented:
Yeah we can make separate handlers for the pages. Do I make separate handlers for getting the explorations also? Like currently I have `LibraryGroupIndexHandler` which fetches the explorations for top rated and recently published. Do I make separate handlers for this task also?

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -73,6 +73,12 @@ class LibraryPage(base.BaseHandler):
     def get(self):
         """Handles GET requests."""
         search_mode = 'search' in self.request.url
+        rank_mode = 'explorations' in self.request.url
```
Removed

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -85,6 +91,8 @@ def get(self):
             'LANGUAGE_CODES_AND_NAMES': (
                 utils.get_all_language_codes_and_names()),
             'search_mode': search_mode,
+            'rank_mode': rank_mode,
```
Removed

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -99,10 +107,12 @@ def get(self):
             feconf.DEFAULT_LANGUAGE_CODE])
         recently_published_summary_dicts = (
-            summary_services.get_recently_published_exploration_summary_dicts())
+            summary_services.get_recently_published_exploration_summary_dicts( # pylint: disable=line-too-long
```
Removed

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -115,24 +125,27 @@ def get(self):
             summary_dicts_by_category.insert(0, {
                 'activity_summary_dicts': recently_published_summary_dicts,
-                'categories': [],
+                'categories': ['recently_published'],
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
 
 
+class LibraryGroupPage(base.BaseHandler):
+    """The page for displaying top rated and recently published explorations
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+    def get(self, rank_method):
+        """Handles GET requests."""
+
+        search_mode = 'search' in self.request.url
```
Removed

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+        search_mode = 'search' in self.request.url
+        rank_mode = 'explorations' in self.request.url
+
+        if rank_method == 'recently_published':
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+
+
+class LibraryGroupIndexHandler(base.BaseHandler):
+    """Provides data for categories such as top rated and recently published"""
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+class LibraryGroupIndexHandler(base.BaseHandler):
+    """Provides data for categories such as top rated and recently published"""
+
+    def get(self, rank_method):
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+
+        if rank_method == 'recently_published':
+            recently_published_summary_dicts = (
+                summary_services.get_recently_published_exploration_summary_dicts( # pylint: disable=line-too-long
```
Done

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+            if recently_published_summary_dicts:
+                summary_dicts_by_category.insert(0, {
+                    'activity_summary_dicts': recently_published_summary_dicts,
+                    'categories': ['recently_published'],
```
Done

Arunabh98 commented on main.py:
```diff
@@ -195,7 +195,11 @@ def ui_access_wrapper(self, *args, **kwargs):
         resources.ValueGeneratorHandler),
 
     get_redirect_route(r'%s' % feconf.LIBRARY_INDEX_URL, library.LibraryPage),
+    get_redirect_route(r'/explorations/<rank_method>',
```
Done except that instead of <group_name> has been split into two separate urls.

Arunabh98 commented on main.py:
```diff
@@ -195,7 +195,11 @@ def ui_access_wrapper(self, *args, **kwargs):
+    get_redirect_route(r'/explorations/<rank_method>',
+                       library.LibraryGroupPage),
     get_redirect_route(r'/libraryindexhandler', library.LibraryIndexHandler),
+    get_redirect_route(r'/librarygroupindexhandler/<rank_method>',
```
Done

seanlip commented:
I think one handler is enough in that case, because you can just pass the
group name in as one of the keys in the request's JSON payload.

On Tuesday, September 27, 2016, Arunabh <notifications@github.com> wrote:

> Yeah we can make separate handlers for the pages. Do I make separate
> handlers for getting the explorations also? Like currently I have
> LibraryGroupIndexHandler which fetches the explorations for top rated and
> recently published. Do I make separate handlers for this task also?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2507#issuecomment-249905373>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeynEGF05bXmaPpm1Kh60U-mHJ1leCks5quTmegaJpZM4KErIl>
> .
>

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +156,87 @@ def get(self):
+    def get(self, rank_method):
+        summary_dicts_by_category = []
+
+        if rank_method == 'recently_published':
```
Done

Arunabh98 commented:
PTAL

seanlip commented on feconf.py:
```diff
@@ -753,6 +767,10 @@ def get_empty_ratings():
     'Engage with the Oppia community by discussing questions, bugs and '
     'explorations in the forum.')
+GROUP_PAGE_DESCRIPTION = (
+    'Discover the explorations belonging to a particular group such as '
```
Discover top-rated or recently-published explorations on Oppia. Learn from ...

Arunabh98 replied:
Done

seanlip commented:
Thanks! Made additional code suggestions. Some additional comments, too:
- There seem to be no tests for the new handlers?
- When I load no explorations and go to /library/top_rated, I see just a blank page. Shouldn't there be some text along the lines of "There are no rated explorations to display" or something like that?

Also, I'm doing a UI test now and may comment more in a bit.

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +148,83 @@ def get(self):
+
+    def get(self):
+        """Handles GET requests."""
+        group_mode = (feconf.LIBRARY_GROUP_RECENTLY_PUBLISHED or
```
Yeah I agree. I will set group_mode = True. Just introduced the check to be on the safe side.

Arunabh98 replied:
Done

seanlip commented:
OK, only one other thought on playing with the UI. Might it be a good idea to have a header like "Top-Rated Explorations" or similar on the library group pages? Otherwise, I just see a bunch of cards, but there's no other indication about what the page itself represents.

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +148,83 @@ def get(self):
+                    'header': feconf.LIBRARY_CATEGORY_RECENTLY_PUBLISHED,
+                    'has_full_results_page': True,
+                    'full_results_url': (
+                        '/library/' + feconf.LIBRARY_GROUP_RECENTLY_PUBLISHED),
```
Agreed. Will be removed

Arunabh98 replied:
Done!

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -118,6 +120,7 @@ def get(self):
                 'categories': [],
                 'header_i18n_id': feconf.LIBRARY_CATEGORY_RECENTLY_PUBLISHED,
                 'has_full_results_page': True,
+                'full_results_url': '/library/recently_published',
```
Done!

Arunabh98 commented on core/controllers/library.py:
```diff
@@ -143,6 +148,83 @@ def get(self):
+
+        self.values.update({
+            'meta_description': (
+                feconf.GROUP_PAGE_DESCRIPTION),
```
Done!

Arunabh98 commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +40,56 @@ oppia.controller('Library', [
-        $scope.leftmostCardIndices.push(0);
-      }
-    });
+    $scope.inRankMode = ($window.location.pathname.indexOf(
```
Modified the if condition to:
```
$scope.inGroupMode = ($window.location.pathname.indexOf(
      'recently_published') === 9) || ($window.location.pathname.indexOf(
      'top_rated') === 9);
```
which holds for only recently published and top rated pages.

Arunabh98 commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -3,6 +3,8 @@
 {% block maintitle %}
   {% if search_mode %}
```
Done

Arunabh98 commented:
1) Added a header similar to the ones in library group page.
2) Added tests for the handlers
3) Added message to display when there are no explorations to display.

Arunabh98 commented:
PTAL!

seanlip commented on core/controllers/library.py:
```diff
@@ -143,6 +148,78 @@ def get(self):
+
+    def get(self):
+        """Handles GET requests."""
+        group_mode = True
```
No need to declare this separately, just substitute group_mode for True in the return dict.

Arunabh98 replied:
Removed group_mode

seanlip commented on core/controllers/library_test.py:
```diff
@@ -44,6 +45,59 @@ def test_library_page(self):
         self.assertEqual(response.status_int, 200)
         response.mustcontain('I18N_LIBRARY_PAGE_TITLE')
 
+    def test_group_pages(self):
```
group --> library_group

seanlip replied:
Can you also break this and the following test out into a separate test class?

Arunabh98 replied:
Done!

seanlip commented on core/controllers/library_test.py:
```diff
@@ -44,6 +45,59 @@ def test_library_page(self):
+        response = self.testapp.get(feconf.LIBRARY_RECENTLY_PUBLISHED_URL)
+        self.assertEqual(response.status_int, 200)
+
+    def test_library_handler_for_group_pages(self):
```
test_handler_for_library_group_pages

Arunabh98 replied:
Done!

seanlip commented on core/controllers/library_test.py:
```diff
@@ -44,6 +45,59 @@ def test_library_page(self):
+            'activity_summary_dicts_by_category': [],
+            'preferred_language_codes': ['en'],
+            'profile_picture_data_url': None,
+            }, response_dict)
```
deindent by 4 to match opening brace

Arunabh98 replied:
Done

seanlip commented on core/controllers/library_test.py:
```diff
@@ -44,6 +45,59 @@ def test_library_page(self):
+            'language_code': 'en',
+            'objective': 'become familiar with Oppia\'s capabilities',
+            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
+        }, response_dict['activity_summary_dicts_by_category'][0]['activity_summary_dicts'][0]) # pylint: disable=line-too-long
```
So, looking at this, I'm actually wondering why we're bothering with the outer two levels. Couldn't we just return a simpler dict with keys 'activity_summary_dicts', 'header_i18n_id' and 'preferred_language_codes' instead?

Arunabh98 replied:
Could you explain this a bit more? LibraryGroupIndexHandler returns -
```
            'activity_summary_dicts_by_category': (
                summary_dicts_by_category),
            'preferred_language_codes': preferred_language_codes,
```
which is same as what is in the same format as LibraryIndexHandler. So what exactly are you suggesting?

Arunabh98 replied:
Yes, that would be more efficient. Will do this...

Arunabh98 replied:
Done

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +40,57 @@ oppia.controller('Library', [
-        $scope.leftmostCardIndices.push(0);
-      }
-    });
+    $scope.inGroupMode = ($window.location.pathname.indexOf(
```
Well... this is kind of messy, tbh, and fragile to changes in the url. How about defining GLOBALS.GROUP_MODE and using that here? (See the header_js block in the html file.)

Arunabh98 replied:
Used pageMode

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +40,57 @@ oppia.controller('Library', [
+      'recently_published') === 9) || ($window.location.pathname.indexOf(
+      'top_rated') === 9);
+    $scope.groupName = $window.location.pathname.substr(
+      $window.location.pathname.lastIndexOf('/') + 1);
```
I think it's more robust to either:
(a) take the pathname, split by '/' and look at the correspondingly-numbered entry, as is done in explorationContextService.js
(b) pass group_name from the backend and set GLOBALS.GROUP_NAME.

The issue with doing it the current way is that it's not robust in cases where we might add, e.g. a url hash at the end that has a '/' after it. On the other hand, the base pathname is never going to change.

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -198,15 +221,19 @@ oppia.controller('Library', [
-      for (i = 0; i < categories.length; i++) {
-        selectedCategories[categories[i]] = true;
-      }
+    $scope.showAllResultsForCategories = function(categories, fullResultsUrl) {
```
Maybe rename to showFullResultsPage, since some of these pages don't correspond to categories.

Also add docstring to explain what this function really does (and, in particular, that if fullResultsUrl is supplied, categories is ignored). That's not going to be obvious to a casual reader.

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -43,7 +43,39 @@
 
-        <div class="oppia-exp-summary-tiles-container" ng-if="!inSearchMode">
+        <div class="oppia-exp-summary-tiles-container" ng-if="inGroupMode" style="margin-bottom: 20px;">
+          <div ng-if="libraryGroups.length == 0"
```
===

Arunabh98 replied:
Done

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -43,7 +43,39 @@
+          </div>
+
+          <div ng-if="libraryGroups.length > 0"
+               ng-repeat="group in libraryGroups track by $index"
```
Related to a previous comment, I'm not sure why we bother with ng-repeat here. There's only going to be one group on this page...

Arunabh98 replied:
Done

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -43,7 +43,39 @@
+          </div>
+        </div>
+
+        <div class="oppia-exp-summary-tiles-container" ng-if="!inSearchMode && !inGroupMode">
```
One thought: instead of having two boolean flags, wouldn't it be better to have a single mode flag that took the values LIBRARY_MODE_GROUP, LIBRARY_MODE_SEARCH, LIBRARY_MODE_INDEX? These constants should be defined both in feconf and in the frontend (the following is the frontend version):

```
    oppia.constant('LIBRARY_PAGE_MODES', {
      GROUP: ...,
      INDEX: ...,
      SEARCH: ...
    })
```

and comments should be added to each to say "make sure this is in sync with the other one".

Arunabh98 replied:
Could you please explain this a bit more? Like how exactly would I use LIBRARY_PAGE_MODES? Also what would be the values of GROUP and LIBRARY_MODE_GROUP? 

This is what I understood:
I'll make a function called getPageMode. Then the condition in ng-if would be something like this `getPageMode() === LIBRARY_PAGE_MODES.INDEX`. Is this what you had in mind?

seanlip replied:
Nope, but close. What I meant was, have a $scope.pageMode = ... (where "..." is passed from the backend). Then do comparisons like ng-if="pageMode === LIBRARY_PAGE_MODES.INDEX".

In the frontend, you have:

```
    oppia.constant('LIBRARY_PAGE_MODES', {
      GROUP: 'group',
      INDEX: 'index',
      SEARCH: 'search'
    })
```

In the backend, you have:

```
    LIBRARY_PAGE_MODE_GROUP = 'group'
    LIBRARY_PAGE_MODE_INDEX = 'index'
    LIBRARY_PAGE_MODE_SEARCH = 'search'
```

Arunabh98 replied:
Should I still leave  inSearchMode for `LibraryFooter.js` as it is or should I modify this also?

seanlip replied:
Good question. I think, modify it too, so that we have consistency.

Arunabh98 replied:
Okay. Also is it okay to have a condition like this in `library.html` -
```
{% if page_mode == 'group' or page_mode == 'search' %}
    Find explorations to learn from - Oppia
```
I am not able to figure out a way to have a condition like this:
```
{% if page_mode == LIBRARY_PAGE_MODES.GROUP  or page_mode == LIBRARY_PAGE_MODES.SEARCH %}
    Find explorations to learn from - Oppia
```

seanlip replied:
Yes, that's fine. We'll fix that part later.

Also, maybe you can do: `if page_mode in ['group', 'search']` -- it's a bit more concise.

Arunabh98 replied:
Done

seanlip commented on feconf.py:
```diff
@@ -445,10 +453,13 @@ def get_empty_ratings():
 FEEDBACK_THREADLIST_URL_PREFIX = '/threadlisthandler'
 FEEDBACK_THREAD_VIEW_EVENT_URL = '/feedbackhandler/thread_view_event'
 FLAG_EXPLORATION_URL_PREFIX = '/flagexplorationhandler'
+LIBRARY_GROUP_INDEX_URL = '/librarygroupindexhandler'
```
LIBRARY_GROUP_DATA_URL

seanlip replied:
and use /librarygrouphandler. The 'index' actually refers to the 'index page', but this isn't an index page.

Arunabh98 replied:
Done!

seanlip commented on main.py:
```diff
@@ -195,7 +195,14 @@ def ui_access_wrapper(self, *args, **kwargs):
+                       library.LibraryGroupPage),
+    get_redirect_route(r'%s' % feconf.LIBRARY_TOP_RATED_URL,
+                       library.LibraryGroupPage),
+    get_redirect_route(r'%s' % feconf.LIBRARY_INDEX_DATA_URL,
```
Keep this one below LIBRARY_INDEX_URL; they go together.

seanlip commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -43,7 +43,39 @@
+          <div ng-if="libraryGroups.length == 0"
+               style="font-size: 1.2em; padding-bottom: 40px; padding-top: 40px;">
+              <p style="text-align: center;">
+                <span>Oops, there are no explorations to display.</span>
```
This message should be internationalized.

Arunabh98 replied:
How do I translate this string in other languages? I do not know it's translation in other languages.

seanlip replied:
Use Google translate to start with (and sanity-check by comparing to similar messages in the i18n json file).

Arunabh98 replied:
I thought about using google translate but was worried about the accuracy. Just wanted to make sure that it's ok to use Google translate.

Will do this...

Arunabh98 replied:
Done!

Arunabh98 commented:
PTAL!

seanlip commented on assets/i18n/id.json:
```diff
@@ -142,6 +142,7 @@
   "I18N_LIBRARY_LOADING": "Memuat",
   "I18N_LIBRARY_MAIN_HEADER": "Bayangkan apa yang dapat anda pelajari hari ini...",
   "I18N_LIBRARY_N/A": "N/A",
+  "I18N_LIBRARY_NO_EXPLORATIONS": "Ups, tidak ada eksplorasi untuk ditampilkan.",
```
untuk --> yang dapat

Arunabh98 replied:
Done!

seanlip commented on assets/i18n/pt.json:
```diff
@@ -142,6 +142,7 @@
   "I18N_LIBRARY_LOADING": "A carregar",
   "I18N_LIBRARY_MAIN_HEADER": "Imagina o que poderás aprender hoje...",
   "I18N_LIBRARY_N/A": "N/D",
+  "I18N_LIBRARY_NO_EXPLORATIONS": "Oops, não existem explorações para mostrar.",
```
Oops, não há explorações para apresentar.

Arunabh98 replied:
Done!

seanlip commented on assets/i18n/qqq.json:
```diff
@@ -142,6 +142,7 @@
   "I18N_LIBRARY_LOADING": "Text displayed in the middle of the page while the content is being loaded.",
   "I18N_LIBRARY_MAIN_HEADER": "Main title of the library page. - The library page shows all the explorations and collections available. It should be a single phrase and fit in one line.",
   "I18N_LIBRARY_N/A": "Text displayed inside the exploration card in the library - It's shown when the exploration has no rating. It should be very small as it's replacing a two digit number.",
+  "I18N_LIBRARY_NO_EXPLORATIONS": "Text displayed in the center of the page - It's shown when there are no explorations to display on a group page such as top rated or recently published explorations.",
```
Wording suggestions:
- center of the **library group** page
- to display on a **library** group page

Arunabh98 replied:
Done!

seanlip commented on assets/i18n/vi.json:
```diff
@@ -142,6 +142,7 @@
   "I18N_LIBRARY_LOADING": "Đang chạy",
   "I18N_LIBRARY_MAIN_HEADER": "Hãy tưởng tượng những gì bạn có thể học hôm nay....",
   "I18N_LIBRARY_N/A": "Không có",
+  "I18N_LIBRARY_NO_EXPLORATIONS": "Rất tiếc, không có cuộc thám hiểm để hiển thị.",
```
Hi @MinhHuong, is this correct? (It's meant to be a translation of "Oops, there are no explorations to display.").

Arunabh98 replied:
Leaving it the same for now!

MinhHuong replied:
I think it's pretty good already

seanlip commented on core/controllers/library_test.py:
```diff
@@ -187,6 +188,62 @@ def test_library_handler_for_created_explorations(self):
+        response_dict = self.get_json(
+            feconf.LIBRARY_GROUP_DATA_URL,
+            {'group_name': feconf.LIBRARY_GROUP_RECENTLY_PUBLISHED})
+        self.assertDictContainsSubset({
```
Why not test all of response_dict?

Arunabh98 replied:
Would having two tests in this order suffice:
```
        self.assertDictContainsSubset({
            'header_i18n_id': 'I18N_LIBRARY_GROUPS_RECENTLY_PUBLISHED',
            'preferred_language_codes': ['en'],
        }, response_dict)
        self.assertDictContainsSubset({
            'id': '0',
            'category': 'Welcome',
            'title': 'Welcome to Oppia!',
            'language_code': 'en',
            'objective': 'become familiar with Oppia\'s capabilities',
            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
        }, response_dict['activity_list'][0])
```
Testing activity_list along with header_i18n_id and preferred_language_codes is difficult as I only want to check if activity_list has a subset of the keys. activity_list has many other keys such as `'created_on_msec': 1475403322644.269` which we do not want to check.

seanlip replied:
That's fine; just add another check that the activity_list has length 1.

Arunabh98 replied:
Done!

seanlip commented on core/controllers/library_test.py:
```diff
@@ -187,6 +188,62 @@ def test_library_handler_for_created_explorations(self):
+        response_dict = self.get_json(
+            feconf.LIBRARY_GROUP_DATA_URL,
+            {'group_name': feconf.LIBRARY_GROUP_TOP_RATED})
+        self.assertDictContainsSubset({
```
Ditto: test the whole thing, not just a part of it.

Arunabh98 replied:
Done!

seanlip commented on core/controllers/library_test.py:
```diff
@@ -187,6 +188,62 @@ def test_library_handler_for_created_explorations(self):
+            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
+        }, response_dict['activity_list'][0])
+
+        # Assign rating to exploration to test handler for top rated
```
Split this into another test. Load the demo exploration and test that the response has 0 explorations. Then give it a rating and test that the response_dict has one exploration.

Then load another exploration, give it a higher rating, and test that the response_dict has two explorations and that they come out in the right order.

Arunabh98 replied:
Is this what you had in mind?
```
        # Load a public demo exploration.
        exp_services.load_demo('0')

        response_dict = self.get_json(
            feconf.LIBRARY_GROUP_DATA_URL,
            {'group_name': feconf.LIBRARY_GROUP_TOP_RATED})
        self.assertDictContainsSubset({
            'is_admin': False,
            'is_moderator': False,
            'is_super_admin': False,
            'activity_list': [],
            'preferred_language_codes': ['en'],
            'profile_picture_data_url': None,
        }, response_dict)

        # Assign rating to exploration to test handler for top rated
        # explorations page.
        rating_services.assign_rating_to_exploration('user', '0', 2)

        # Test whether the response contains the exploration we have rated.
        response_dict = self.get_json(
            feconf.LIBRARY_GROUP_DATA_URL,
            {'group_name': feconf.LIBRARY_GROUP_TOP_RATED})
        self.assertEqual(len(response_dict['activity_list']), 1)
        self.assertDictContainsSubset({
            'id': '0',
            'category': 'Welcome',
            'title': 'Welcome to Oppia!',
            'language_code': 'en',
            'objective': 'become familiar with Oppia\'s capabilities',
            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
        }, response_dict['activity_list'][0])

        # Load another public demo exploration.
        exp_services.load_demo('1')

        # Assign rating to exploration to test handler for top rated
        # explorations page.
        rating_services.assign_rating_to_exploration('user', '1', 4)

        # Test whether the response contains both the explorations we have
        # rated and they are returned in decending order of rating.
        response_dict = self.get_json(
            feconf.LIBRARY_GROUP_DATA_URL,
            {'group_name': feconf.LIBRARY_GROUP_TOP_RATED})
        self.assertEqual(len(response_dict['activity_list']), 2)
        self.assertDictContainsSubset({
            'id': '1',
            'category': 'Programming',
            'title': 'Project Euler Problem 1',
            'language_code': 'en',
            'objective': 'solve Problem 1 on the Project Euler site',
            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
        }, response_dict['activity_list'][0])
        self.assertDictContainsSubset({
            'id': '0',
            'category': 'Welcome',
            'title': 'Welcome to Oppia!',
            'language_code': 'en',
            'objective': 'become familiar with Oppia\'s capabilities',
            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
        }, response_dict['activity_list'][1])

seanlip replied:
Yes, thanks -- this is good! I would also suggest adding a single line somewhere to check that the response_dict has the correct header_i18n_id.

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -16,13 +16,23 @@
  * @fileoverview Data and controllers for the Oppia contributors' library page.
  */
 
+// The constants defined below in LIBRARY_PAGE_MODES should be same as the
```
This message is great. Can you prefix it with "NOTE TO DEVELOPERS: "?

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +52,56 @@ oppia.controller('Library', [
-
-      $rootScope.$broadcast(
-        'preferredLanguageCodesLoaded', data.preferred_language_codes);
+    var pathnameArray = $window.location.pathname.split('/');
```
Only do this check within the if block.

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +52,56 @@ oppia.controller('Library', [
-      $timeout(function() {
-        tileDisplayWidth = $('exploration-summary-tile').width() || 214;
-      }, 20);
+        $scope.groupHeader = data.header_i18n_id;
```
groupHeaderI18nId?

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -40,34 +52,56 @@ oppia.controller('Library', [
+      });
+    } else {
+      $http.get('/libraryindexhandler').success(
+      function(data) {
```
I think you can move this to the end of the previous line?

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/Library.js:
```diff
@@ -185,28 +219,33 @@ oppia.controller('Library', [
-      }
+    // The following loads explorations belonging to a particular group. If
+    // fullResultsUrl is given it loads the page corresponding to the url. If
+    // not given it will initiate a search query based on categories.
```
"If not given " --> "Otherwise, "
"based on categories" --> "for the given list of categories."

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/LibraryFooter.js:
```diff
@@ -16,8 +16,20 @@
  * @fileoverview Data and controllers for the Oppia library footer.
  */
 
+// The constants defined below in LIBRARY_PAGE_MODES should be same as the
```
Actually, drop this and modify the other comments to exclude reference to LibraryFooter.js. I forgot that all frontend files will get concatenated, so we don't really need to declare this twice with one overwriting the other. Sorry!

Arunabh98 replied:
Done!

seanlip commented on core/templates/dev/head/pages/library/LibraryFooter.js:
```diff
@@ -16,8 +16,20 @@
-  $scope.inSearchMode = ($window.location.pathname.indexOf('/search') === 0);
-}]);
+  '$scope', '$window', 'LIBRARY_PAGE_MODES', function(
+      $scope, $window, LIBRARY_PAGE_MODES) {
```
Don't need $window any more?

Arunabh98 replied:
Done!

Arunabh98 commented:
PTAL!

seanlip commented on core/templates/dev/head/pages/library/LibraryFooter.js:
```diff
@@ -16,8 +16,15 @@
  * @fileoverview Data and controllers for the Oppia library footer.
  */
 
+oppia.constant('LIBRARY_PAGE_MODES', {
```
I thought we were getting rid of this? It's already part of Library.js.

seanlip commented:
LGTM, thanks @Arunabh98! We can merge when Travis checks pass.
